Sources/FirebaseFirestore/vendor/Codable/DocumentID.swift

```diff
@@ -0,0 +1,181 @@
+  public static func wrap(_ documentReference: DocumentReference) throws -> Self {
+    return documentReference.documentID
+  }
+}
+
+extension DocumentReference: DocumentIDWrappable {
+  public static func wrap(_ documentReference: DocumentReference) throws -> Self {
+    // Swift complains that values of type DocumentReference cannot be returned
+    // as Self which is nonsensical. The cast forces this to work.
+    return documentReference as! Self
+  }
+}
+
+/// An internal protocol that allows Firestore.Decoder to test if a type is a
+/// DocumentID of some kind without knowing the specific generic parameter that
+/// the user actually used.
+///
+/// This is required because Swift does not define an existential type for all
+/// instances of a generic class--that is, it has no wildcard or raw type that
+/// matches a generic without any specific parameter. Swift does define an
+/// existential type for protocols though, so this protocol (to which DocumentID
+/// conforms) indirectly makes it possible to test for and act on any
+/// `DocumentID<Value>`.
+internal protocol DocumentIDProtocol {
+  /// Initializes the DocumentID from a DocumentReference.
+  init(from documentReference: DocumentReference?) throws
+}
+
+/// A property wrapper type that marks a `DocumentReference?` or `String?` field to
+/// be populated with a document identifier when it is read.
+///
+/// Apply the `@DocumentID` annotation to a `DocumentReference?` or `String?`
+/// property in a `Codable` object to have it populated with the document
+/// identifier when it is read and decoded from Firestore.
+///
+/// - Important: The name of the property annotated with `@DocumentID` must not
+///   match the name of any fields in the Firestore document being read or else
+///   an error will be thrown. For example, if the `Codable` object has a
+///   property named `firstName` annotated with `@DocumentID`, and the Firestore
+///   document contains a field named `firstName`, an error will be thrown when
+///   attempting to decode the document.
+///
+/// - Example Read:
+///   ````
+///   struct Player: Codable {
+///     @DocumentID var playerID: String?
+///     var health: Int64
+///   }
+///
+///   let p = try! await Firestore.firestore()
+///       .collection("players")
+///       .document("player-1")
+///       .getDocument(as: Player.self)
+///   print("\(p.playerID!) Health: \(p.health)")
+///
+///   // Prints: "Player: player-1, Health: 95"
+///   ````
+///
+/// - Important: Trying to encode/decode this type using encoders/decoders other than
+///   Firestore.Encoder throws an error.
+///
+/// - Important: When writing a Codable object containing an `@DocumentID` annotated field,
+///   its value is ignored. This allows you to read a document from one path and
+///   write it into another without adjusting the value here.
+@propertyWrapper
+public struct DocumentID<Value: DocumentIDWrappable & Codable>:
+  StructureCodingUncodedUnkeyed {
```

Comment:
Could you please indent the continuation an extra level to discern it from the indented body.